dev-lead: git worktree collision ('branch already used by worktree') breaks review-changes — blocks PR #383

## Summary

Dev-lead `review-changes`/`fix-reviews` runs fail with a git worktree collision when the target branch is already checked out (or claimed by a stale worktree) in the runner workspace:

```
Previous HEAD position was e2a1e68 Merge eb5244b... into d4debee...
Switched to a new branch 'dev-lead/issue-382-20260524-0133'
branch 'dev-lead/issue-382-20260524-0133' set up to track 'origin/dev-lead/issue-382-20260524-0133'.
Preparing worktree (detached HEAD eb5244b)
fatal: 'dev-lead/issue-382-20260524-0133' is already used by worktree at '/tmp/dev-lead-wt-IOkyR8/pr-383'
failed to run git: exit status 128
##[error]Process completed with exit code 1.
```

Observed on run [27107230249](https://github.com/petry-projects/.github-private/actions/runs/27107230249/job/79998466680) (Dev-Lead Agent, `review-changes` for PR #383, 2026-06-07). The failure is reproducible: a rerun of the same job failed identically.

## Impact

- The failed `dispatch` check attaches to PR #383's head SHA → the PR's check rollup shows `failing` → the PR-review cascade correctly skips it (`ci-failing`) → **PR #383 can never get an automated review** until this is fixed.
- Any dev-lead intent that prepares a worktree for a branch that was just checked out in the main workspace will hit this.

## Root cause (suspected)

The job first does a plain `git checkout <branch>` (making the primary workspace claim the branch), then runs `git worktree add` (via `scripts/lib/pr-worktree.sh`) for the same branch at `/tmp/dev-lead-wt-*/pr-<n>` — git refuses because a branch can only be checked out in one worktree at a time. A stale worktree registration from a previous step in the same run produces the same error.

## Proposed fix (for dev-lead)

In `scripts/lib/pr-worktree.sh` (or the dev-lead checkout step):

1. Before `git worktree add`, run `git worktree prune` and, if the branch is already checked out elsewhere, either reuse that checkout or detach the primary workspace first (`git checkout --detach`).
2. Alternatively, create the worktree with `--force` only after verifying the existing registration is stale, or use a detached worktree (`git worktree add --detach` + `git switch -C`) so branch ownership never conflicts.
3. Add a bats test covering: branch checked out in primary workspace, stale worktree entry, and the happy path.

## Acceptance criteria

- `review-changes` on PR #383 completes without `exit status 128`.
- Re-running a dev-lead job after a partial failure does not collide with leftover worktrees.

--- a/scripts/lib/pr-worktree.sh
+++ b/scripts/lib/pr-worktree.sh
@@ -85,14 +85,16 @@ checkout_pr_in_worktree() {
   local _pr_branch=""
   _pr_branch="$(gh pr view "$pr" --repo "$repo" --json headRefName --jq '.headRefName' 2>/dev/null || true)"
   if [ -n "$_pr_branch" ]; then
-    # Canonicalize the agent dir once for comparison: `git worktree list`
-    # emits fully-resolved paths, but PR_WORKTREE_AGENT_DIR comes from `pwd`
-    # and may contain unresolved symlinks (macOS, some CI runners). Without
-    # this, the agent's own checkout could fail the equality test and be sent
-    # down the force-remove path (a no-op git refuses, leaving the branch
-    # unreleased and the original collision unfixed).
+    # Resolve the agent worktree's ROOT for comparison. `git worktree list`
+    # emits each worktree's top-level, fully-resolved path, but
+    # PR_WORKTREE_AGENT_DIR comes from `pwd` — which may be a SUBDIRECTORY of
+    # the repo and/or contain unresolved symlinks (macOS, some CI runners).
+    # Comparing the resolved repo top-level against git's output on both axes
+    # ensures the agent's own checkout is detached (not mistakenly force-removed,
+    # which git refuses — leaving the branch unreleased and the collision unfixed).
     local _agent_real
-    _agent_real="$(cd "$PR_WORKTREE_AGENT_DIR" 2>/dev/null && pwd -P || printf '%s' "$PR_WORKTREE_AGENT_DIR")"
+    _agent_real="$(git -C "$PR_WORKTREE_AGENT_DIR" rev-parse --show-toplevel 2>/dev/null || printf '%s' "$PR_WORKTREE_AGENT_DIR")"
+    _agent_real="$(cd "$_agent_real" 2>/dev/null && pwd -P || printf '%s' "$_agent_real")"
     local _wt_line _wt_path="" _held _wt_real
     while IFS= read -r _wt_line; do
       case "$_wt_line" in

--- a/scripts/lib/pr-worktree.sh
+++ b/scripts/lib/pr-worktree.sh
@@ -70,6 +70,42 @@ checkout_pr_in_worktree() {
   # shellcheck disable=SC2064
   trap "cleanup_pr_worktree${_prev_exit:+; $_prev_exit}" EXIT
 
+  # Clear stale worktree state left by crashed prior runs on this same runner
+  # before we check the PR branch out. Without this, `gh pr checkout` below
+  # fails hard with
+  #   fatal: '<branch>' is already used by worktree at '<stale path>'
+  # because a previous run left the PR's head branch checked out in a worktree
+  # that is now defunct (issue #470, run 27107230249 on PR #383).
+  #
+  #   1. `git worktree prune` drops registry entries whose directories are gone.
+  #   2. For entries whose directory still exists, look up the PR's head branch
+  #      and release whatever worktree currently holds it: detach the main agent
+  #      checkout in place (it cannot be removed), force-remove any other.
+  git worktree prune 2>/dev/null || true
+  local _pr_branch=""
+  _pr_branch="$(gh pr view "$pr" --repo "$repo" --json headRefName --jq '.headRefName' 2>/dev/null || true)"
+  if [ -n "$_pr_branch" ]; then
+    local _wt_line _wt_path="" _held
+    while IFS= read -r _wt_line; do
+      case "$_wt_line" in
+        "worktree "*) _wt_path="${_wt_line#worktree }" ;;
+        "branch refs/heads/"*)
+          _held="${_wt_line#branch refs/heads/}"
+          if [ "$_held" = "$_pr_branch" ] && [ -n "$_wt_path" ]; then
+            if [ "$_wt_path" = "$PR_WORKTREE_AGENT_DIR" ]; then
+              # The agent's own checkout holds the branch — never remove it;
+              # detaching releases the branch so the new worktree can claim it.
+              git -C "$_wt_path" checkout --detach --quiet 2>/dev/null || true
+            else
+              git worktree remove --force "$_wt_path" 2>/dev/null || true
+            fi
+          fi
+          ;;
+      esac
+    done < <(git worktree list --porcelain 2>/dev/null)
+    git worktree prune 2>/dev/null || true
+  fi
+
   # Start detached at the agent checkout's HEAD so `gh pr checkout` can create or
   # switch to the PR branch inside the worktree without colliding with whatever
   # branch the agent checkout already has out.

--- a/scripts/lib/pr-worktree.sh
+++ b/scripts/lib/pr-worktree.sh
@@ -85,14 +85,23 @@ checkout_pr_in_worktree() {
   local _pr_branch=""
   _pr_branch="$(gh pr view "$pr" --repo "$repo" --json headRefName --jq '.headRefName' 2>/dev/null || true)"
   if [ -n "$_pr_branch" ]; then
-    local _wt_line _wt_path="" _held
+    # Canonicalize the agent dir once for comparison: `git worktree list`
+    # emits fully-resolved paths, but PR_WORKTREE_AGENT_DIR comes from `pwd`
+    # and may contain unresolved symlinks (macOS, some CI runners). Without
+    # this, the agent's own checkout could fail the equality test and be sent
+    # down the force-remove path (a no-op git refuses, leaving the branch
+    # unreleased and the original collision unfixed).
+    local _agent_real
+    _agent_real="$(cd "$PR_WORKTREE_AGENT_DIR" 2>/dev/null && pwd -P || printf '%s' "$PR_WORKTREE_AGENT_DIR")"
+    local _wt_line _wt_path="" _held _wt_real
     while IFS= read -r _wt_line; do
       case "$_wt_line" in
         "worktree "*) _wt_path="${_wt_line#worktree }" ;;
         "branch refs/heads/"*)
           _held="${_wt_line#branch refs/heads/}"
           if [ "$_held" = "$_pr_branch" ] && [ -n "$_wt_path" ]; then
-            if [ "$_wt_path" = "$PR_WORKTREE_AGENT_DIR" ]; then
+            _wt_real="$(cd "$_wt_path" 2>/dev/null && pwd -P || printf '%s' "$_wt_path")"
+            if [ "$_wt_real" = "$_agent_real" ]; then
               # The agent's own checkout holds the branch — never remove it;
               # detaching releases the branch so the new worktree can claim it.
               git -C "$_wt_path" checkout --detach --quiet 2>/dev/null || true

--- a/tests/dev-lead/unit/test_pr_worktree.bats
+++ b/tests/dev-lead/unit/test_pr_worktree.bats
@@ -13,10 +13,14 @@ setup() {
   STUB_BIN_DIR="$(mktemp -d)"
   # gh stub: `gh pr checkout <n>` switches the worktree to the PR branch,
   # mirroring the real command (which is what made the old prompt files vanish).
+  # `gh pr view --json headRefName --jq .headRefName` returns the PR's head
+  # branch (`old` in this fixture) so the collision-prevention logic in
+  # checkout_pr_in_worktree can identify and release a stale holder (issue #470).
   cat > "$STUB_BIN_DIR/gh" <<'GHEOF'
 #!/usr/bin/env bash
 case "$1 $2" in
   "pr checkout") git checkout -q old ;;
+  "pr view") echo "old" ;;
   *) echo "{}" ;;
 esac
 GHEOF
@@ -123,4 +127,73 @@ teardown() {
 
   cleanup_pr_worktree
   trap - EXIT
+}
+
+# --- issue #470: PR-branch worktree-collision recovery -----------------------
+# `gh pr checkout` fails with "'<branch>' is already used by worktree at ..."
+# when a crashed prior run on the same runner left the PR's head branch claimed.
+# checkout_pr_in_worktree must release the branch before checking it out.
+
+@test "checkout_pr_in_worktree: recovers when PR branch is held by a stale sibling worktree" {
+  source "$LIB"
+  cd "$AGENT_REPO"
+
+  # Simulate a crashed prior run: the PR branch (`old`) is checked out in a
+  # leftover sibling worktree that is still registered AND still on disk.
+  local stale; stale="$(mktemp -d)/stale-wt"
+  git -C "$AGENT_REPO" worktree add --quiet "$stale" old
+  run git -C "$AGENT_REPO" worktree list
+  [[ "$output" == *"$stale"* ]]
+
+  # Without the fix this aborts with "already used by worktree".
+  checkout_pr_in_worktree 383 owner/repo
+
+  [ "$PWD" = "$PR_WORKTREE_DIR" ]
+  # We are on the PR branch in the fresh worktree (prompts/x.md absent on `old`).
+  [ ! -e "$PWD/prompts/x.md" ]
+  # The stale worktree was released.
+  run git -C "$AGENT_REPO" worktree list
+  [[ "$output" != *"$stale"* ]]
+
+  cleanup_pr_worktree
+}
+
+@test "checkout_pr_in_worktree: recovers from a stale registry entry whose dir is gone" {
+  source "$LIB"
+  cd "$AGENT_REPO"
+
+  # Crashed run left a registry entry for `old`, but the directory was deleted
+  # out from under git (no `git worktree remove`). prune must clear it.
+  local stale; stale="$(mktemp -d)/gone-wt"
+  git -C "$AGENT_REPO" worktree add --quiet "$stale" old
+  rm -rf "$stale"
+
+  checkout_pr_in_worktree 383 owner/repo
+
+  [ "$PWD" = "$PR_WORKTREE_DIR" ]
+  [ ! -e "$PWD/prompts/x.md" ]
+
+  cleanup_pr_worktree
+}
+
+@test "checkout_pr_in_worktree: recovers when the agent checkout itself holds the PR branch" {
+  source "$LIB"
+  cd "$AGENT_REPO"
+  # The agent's own checkout is on the PR branch (`old`). It must be detached in
+  # place (never removed) so the new worktree can claim the branch.
+  git -C "$AGENT_REPO" checkout -q old
+
+  checkout_pr_in_worktree 383 owner/repo
+
+  [ "$PWD" = "$PR_WORKTREE_DIR" ]
+  [ ! -e "$PWD/prompts/x.md" ]
+
+  cleanup_pr_worktree
+
+  # The agent checkout still exists and is usable (was detached in place, not
+  # removed) — and it no longer holds the `old` branch, so a future checkout of
+  # the same PR won't collide on it.
+  [ -d "$AGENT_REPO/.git" ]
+  run git -C "$AGENT_REPO" symbolic-ref -q HEAD
+  [ "$status" -ne 0 ]   # detached HEAD has no symbolic ref
 }

--- a/tests/dev-lead/unit/test_pr_worktree.bats
+++ b/tests/dev-lead/unit/test_pr_worktree.bats
@@ -140,7 +140,8 @@ teardown() {
 
   # Simulate a crashed prior run: the PR branch (`old`) is checked out in a
   # leftover sibling worktree that is still registered AND still on disk.
-  local stale; stale="$(mktemp -d)/stale-wt"
+  local stale_parent stale
+  stale_parent="$(mktemp -d)"; stale="$stale_parent/stale-wt"
   git -C "$AGENT_REPO" worktree add --quiet "$stale" old
   run git -C "$AGENT_REPO" worktree list
   [[ "$output" == *"$stale"* ]]
@@ -156,6 +157,7 @@ teardown() {
   [[ "$output" != *"$stale"* ]]
 
   cleanup_pr_worktree
+  rm -rf "$stale_parent"
 }
 
 @test "checkout_pr_in_worktree: recovers from a stale registry entry whose dir is gone" {
@@ -164,7 +166,8 @@ teardown() {
 
   # Crashed run left a registry entry for `old`, but the directory was deleted
   # out from under git (no `git worktree remove`). prune must clear it.
-  local stale; stale="$(mktemp -d)/gone-wt"
+  local stale_parent stale
+  stale_parent="$(mktemp -d)"; stale="$stale_parent/gone-wt"
   git -C "$AGENT_REPO" worktree add --quiet "$stale" old
   rm -rf "$stale"
 
@@ -174,6 +177,7 @@ teardown() {
   [ ! -e "$PWD/prompts/x.md" ]
 
   cleanup_pr_worktree
+  rm -rf "$stale_parent"
 }
 
 @test "checkout_pr_in_worktree: recovers when the agent checkout itself holds the PR branch" {
@@ -196,4 +200,27 @@ teardown() {
   [ -d "$AGENT_REPO/.git" ]
   run git -C "$AGENT_REPO" symbolic-ref -q HEAD
   [ "$status" -ne 0 ]   # detached HEAD has no symbolic ref
+}
+
+@test "checkout_pr_in_worktree: detaches agent checkout even when invoked from a subdirectory" {
+  # `git worktree list` reports the worktree ROOT, but PR_WORKTREE_AGENT_DIR is
+  # captured from pwd — here a SUBDIRECTORY of the repo. The comparison must
+  # resolve to the repo top-level so the agent checkout is still recognized and
+  # detached (regression guard for the PR #494 review finding).
+  source "$LIB"
+  git -C "$AGENT_REPO" checkout -q old
+  mkdir -p "$AGENT_REPO/prompts"   # a real subdir to run from
+  cd "$AGENT_REPO/prompts"
+
+  checkout_pr_in_worktree 383 owner/repo
+
+  [ "$PWD" = "$PR_WORKTREE_DIR" ]
+  [ ! -e "$PWD/prompts/x.md" ]
+
+  cleanup_pr_worktree
+
+  # Agent checkout preserved (detached, not removed despite the subdir cwd).
+  [ -d "$AGENT_REPO/.git" ]
+  run git -C "$AGENT_REPO" symbolic-ref -q HEAD
+  [ "$status" -ne 0 ]
 }

